### PR TITLE
[BEAM-7036] Spark portable runner: stage files to run on 'remote' clu…

### PR DESCRIPTION
**Pull request by ibzib**
…ster

This change lets us run portable pipelines on non-local Spark masters (e.g. spark://host:port). For now, this is just for Java; I haven't tried running any Python yet.

1. Stage files from class path if not manually set in `pipelineOptions`.
2. Move `prepareFilesToStageForRemoteClusterExecution` method from `SparkRunner` to `SparkPipelineOptions` and reuse in portable runner.
3. `prepareFilesToStageForRemoteClusterExecution` should fall back on the JVM's temp dir if not already specified.

R: @angoenka @iemejia 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ibzib** [comment]
Run Python PreCommit

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineRunner.java [hunk 1]
```diff
@@ -53,6 +56,17 @@ public SparkPipelineResult run(RunnerApi.Pipeline pipeline, JobInfo jobInfo) {
...
 
+    if (pipelineOptions.getFilesToStage() == null) {
+      pipelineOptions.setFilesToStage(
+          detectClassPathResourcesToStage(SparkPipelineRunner.class.getClassLoader()));
+      LOG.info(
+          "PipelineOptions.filesToStage was not specified. Defaulting to files from the classpath");
+    }
+    prepareFilesToStageForRemoteClusterExecution(pipelineOptions);
+    LOG.info(
+        "Will stage {} files. Enable logging at DEBUG level to see which files will be staged.");
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/SparkPipelineRunner.java]: note to self: missing argument here

**Comment by angoenka** [comment]
To make it similar to Flink runner, we should use BeamFileSystemArtifactStagingService to stage user code.

**Comment by ibzib** [comment]
> To make it similar to Flink runner, we should use BeamFileSystemArtifactStagingService to stage user code.

Can you point me to where this is used in the Flink runner? I'm not seeing it

**Comment by iemejia** [comment]
This one needs a rebase @ibzib

**Comment by ibzib** [comment]
rebased

**Comment by iemejia** [comment]
Run Portable_Python PreCommit

**Comment by iemejia** [comment]
Run Spark ValidatesRunner

**Pull request merged**